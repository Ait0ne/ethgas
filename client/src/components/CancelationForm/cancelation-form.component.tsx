import React, { useState } from 'react';

import {FormContainer, CustomButton} from '../NotificationForm/notification-form.styles';
import FormInput from '../FormInput/form-input.component';
import {API_URL} from '../../config';

interface CancelationProps {
    toggleAlert: (message:string, success:boolean) => void
}


const CancelationForm: React.FC<CancelationProps> = ({toggleAlert}) => {
    const [email, setEmail] = useState('')
    
    const handleEmailChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setEmail(event.target.value)
    }

    const handleSubmit = (event:React.FormEvent<HTMLFormElement>) => {
        event.preventDefault()
        fetch(`${API_URL}delete`, {
            method: 'POST',
            credentials: 'same-origin',
            headers: {
              'Content-Type': 'application/json'
            },
            body: JSON.stringify({email})
        })
        .then(response => response.json())
        .then(data => {
            if (data.ok) {
                toggleAlert('You successfully unsubscribed', true)
            } else {
                toggleAlert('There was an unexpected error', false)
            }
        })
        .catch(err => {
            toggleAlert('There was an unexpected error', false)
        })
        .finally(() => {
            setEmail('')
        })
    }

    return (
        <FormContainer onSubmit={handleSubmit}>
            <FormInput
            value={email}
            name='email'
            type='email'
            handleChange={handleEmailChange}
            label='Email'
            />
            <CustomButton 
            variant='contained'
            type='submit'
            >
                Unsubscribe
            </CustomButton>
        </FormContainer>
    )
}

export default CancelationForm;
