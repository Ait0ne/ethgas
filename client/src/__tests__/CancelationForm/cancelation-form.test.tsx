import React from 'react';
import {mount} from 'enzyme';

import CancelationForm from '../../components/CancelationForm/cancelation-form.component';
import {CustomButton} from '../../components/NotificationForm/notification-form.styles';

describe('cancelation form', () => {
    const toggleAlert = jest.fn()
    const cancelationForm = mount(<CancelationForm toggleAlert={toggleAlert}/>)


    it('renders correctly', () =>{
        expect(cancelationForm).toMatchSnapshot()
    })

    it('renders one input', () => {
        expect(cancelationForm.find('FormInput').length).toEqual(1)
    })

    it('renders one button', () => {
        expect(cancelationForm.find(CustomButton).length).toEqual(1)
    })


    
    global.fetch = jest.fn(():any =>
    Promise.resolve({
      json: () => Promise.resolve({ ok:true }),
    })
    );

    it('calls fetch on form submit', () => {
        const button =cancelationForm.find(CustomButton)
        button.simulate('submit')

        expect(fetch).toHaveBeenCalledTimes(1)
    })

    it('calls toggle alert function with correct values ', () => {
        expect(toggleAlert).toHaveBeenCalledWith("You successfully unsubscribed", true)
    })

}
)