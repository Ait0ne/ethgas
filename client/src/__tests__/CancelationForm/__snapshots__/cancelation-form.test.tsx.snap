// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cancelation form renders correctly 1`] = `
<CancelationForm
  toggleAlert={[MockFunction]}
>
  <styled.form
    onSubmit={[Function]}
  >
    <form
      className="sc-bdnylx jIfGih"
      onSubmit={[Function]}
    >
      <FormInput
        handleChange={[Function]}
        label="Email"
        name="email"
        type="email"
        value=""
      >
        <styled.div>
          <div
            className="sc-hKFyIo bUmzJB"
          >
            <input
              name="email"
              onChange={[Function]}
              type="email"
              value=""
            />
            <label
              className=""
            >
              Email
            </label>
          </div>
        </styled.div>
      </FormInput>
      <Styled(WithStyles(ForwardRef(Button)))
        type="submit"
        variant="contained"
      >
        <WithStyles(ForwardRef(Button))
          className="sc-dlnjPT juMlrI"
          type="submit"
          variant="contained"
        >
          <ForwardRef(Button)
            className="sc-dlnjPT juMlrI"
            classes={
              Object {
                "colorInherit": "MuiButton-colorInherit",
                "contained": "MuiButton-contained",
                "containedPrimary": "MuiButton-containedPrimary",
                "containedSecondary": "MuiButton-containedSecondary",
                "containedSizeLarge": "MuiButton-containedSizeLarge",
                "containedSizeSmall": "MuiButton-containedSizeSmall",
                "disableElevation": "MuiButton-disableElevation",
                "disabled": "Mui-disabled",
                "endIcon": "MuiButton-endIcon",
                "focusVisible": "Mui-focusVisible",
                "fullWidth": "MuiButton-fullWidth",
                "iconSizeLarge": "MuiButton-iconSizeLarge",
                "iconSizeMedium": "MuiButton-iconSizeMedium",
                "iconSizeSmall": "MuiButton-iconSizeSmall",
                "label": "MuiButton-label",
                "outlined": "MuiButton-outlined",
                "outlinedPrimary": "MuiButton-outlinedPrimary",
                "outlinedSecondary": "MuiButton-outlinedSecondary",
                "outlinedSizeLarge": "MuiButton-outlinedSizeLarge",
                "outlinedSizeSmall": "MuiButton-outlinedSizeSmall",
                "root": "MuiButton-root",
                "sizeLarge": "MuiButton-sizeLarge",
                "sizeSmall": "MuiButton-sizeSmall",
                "startIcon": "MuiButton-startIcon",
                "text": "MuiButton-text",
                "textPrimary": "MuiButton-textPrimary",
                "textSecondary": "MuiButton-textSecondary",
                "textSizeLarge": "MuiButton-textSizeLarge",
                "textSizeSmall": "MuiButton-textSizeSmall",
              }
            }
            type="submit"
            variant="contained"
          >
            <WithStyles(ForwardRef(ButtonBase))
              className="MuiButton-root MuiButton-contained sc-dlnjPT juMlrI"
              component="button"
              disabled={false}
              focusRipple={true}
              focusVisibleClassName="Mui-focusVisible"
              type="submit"
            >
              <ForwardRef(ButtonBase)
                className="MuiButton-root MuiButton-contained sc-dlnjPT juMlrI"
                classes={
                  Object {
                    "disabled": "Mui-disabled",
                    "focusVisible": "Mui-focusVisible",
                    "root": "MuiButtonBase-root",
                  }
                }
                component="button"
                disabled={false}
                focusRipple={true}
                focusVisibleClassName="Mui-focusVisible"
                type="submit"
              >
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-contained sc-dlnjPT juMlrI"
                  disabled={false}
                  onBlur={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="submit"
                >
                  <span
                    className="MuiButton-label"
                  >
                    Unsubscribe
                  </span>
                  <WithStyles(memo)
                    center={false}
                  >
                    <ForwardRef(TouchRipple)
                      center={false}
                      classes={
                        Object {
                          "child": "MuiTouchRipple-child",
                          "childLeaving": "MuiTouchRipple-childLeaving",
                          "childPulsate": "MuiTouchRipple-childPulsate",
                          "ripple": "MuiTouchRipple-ripple",
                          "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                          "rippleVisible": "MuiTouchRipple-rippleVisible",
                          "root": "MuiTouchRipple-root",
                        }
                      }
                    >
                      <span
                        className="MuiTouchRipple-root"
                      >
                        <TransitionGroup
                          childFactory={[Function]}
                          component={null}
                          exit={true}
                        />
                      </span>
                    </ForwardRef(TouchRipple)>
                  </WithStyles(memo)>
                </button>
              </ForwardRef(ButtonBase)>
            </WithStyles(ForwardRef(ButtonBase))>
          </ForwardRef(Button)>
        </WithStyles(ForwardRef(Button))>
      </Styled(WithStyles(ForwardRef(Button)))>
    </form>
  </styled.form>
</CancelationForm>
`;
